### PR TITLE
Make asserting failed exit codes and output nicer

### PR DESCRIPTION
**Pull request by Chemaclass**
## 📚 Description

Closes: https://github.com/TypedDevs/bashunit/issues/334

## 🔖 Changes

- Allow defer calling a function when asserting `exit_code` by passing the callable command to evaluate as raw string instead of `$(...)`
- bashunit output will be render on the stderr

## 🖼️  Demo

Render the stdout in the terminal and the errors in a file

Screenshot 2024-09-28 at 10 21 54

```bash
export PHPSTAN_PATH=[...]/vendor/bin/phpstan

../bashunit -a exit_code "1" "$PHPSTAN_PATH analyze --no-progress --level 8 --error-format raw ./"
echo $?

../bashunit -a exit_code "0" "$PHPSTAN_PATH analyze --no-progress --level 8 --error-format raw ./"
echo $?

../bashunit -a exit_code "0" "$PHPSTAN_PATH analyze --no-progress --level 8 --error-format raw ./" 2> /tmp/error.log
cat /tmp/error.log
```

## ✅ To-do list

- [x] I updated the `CHANGELOG.md` to reflect the new feature or fix
- [x] I updated the documentation to reflect the changes

Screenshot 2024-09-28 at 10 25 43

### CODE REVIEW

File: src/main.sh [hunk 1]
```diff
@@ -34,21 +34,56 @@ function main::exec_tests() {
...
+  # Get the last argument safely by calculating the array length
+  local last_index=$((${#args[@]} - 1))
+  local last_arg="${args[$last_index]}"
+  local inner_exit_code=0
+  local bashunit_exit_code=0
+  local output
+
+  if [[ "$assert_fn" == "assert_exit_code" && "$last_arg" == eval* ]]; then
+    local callable_command="${last_arg#eval }"
+    output=$(eval "$callable_command" 2>&1 || echo "inner_exit_code:$?")
```
Comment (Chemaclass) [inline src/main.sh:60]: `eval` is a real bash command, but in this context of last argument I am using it at `bashunit` level so hide the complexity and bring familiarity with the "bash context" without bringing new weird terms. This is the trick:
```bash
# remove the "eval" word from the last argument
local callable_command="${last_arg#eval }"

# use real bash "eval" to evaluate your command
output=$(eval "$callable_command" 2>&1 || echo "inner_exit_code:$?")
```

File: src/main.sh [hunk 2]
```diff
@@ -34,21 +34,61 @@ function main::exec_tests() {
...
+  local last_arg="${args[$last_index]}"
+  local inner_exit_code=0
+  local bashunit_exit_code=0
+  local output
+
+  if [[ "$assert_fn" == "assert_exit_code" && "$last_arg" == eval* ]]; then
+    # remove the "eval" word from the last argument
+    local callable_command="${last_arg#eval }"
+    # use real bash "eval" to evaluate the command
+    output=$(eval "$callable_command" 2>&1 || echo "inner_exit_code:$?")
```
Comment (Chemaclass) [inline src/main.sh:62]: `eval` is a real bash command, but in this context of last argument I am using it at `bashunit` level to hide the complexity and bring "familiarity" without bringing new terms.

**Comment by ondrejmirtes** [comment]
I have some ideas about this (which you don't have to follow but maybe just think about the direction):

1) If you want `bashunit -a exit_code "1" "eval the_command"` then what would `bashunit -a exit_code "1" "the_command"` mean? Does it mean anything? Can we remove the "eval" and still have it work?

Does -a exit_code imply the thing in double quotes is going to be run as command?

Also trying to think of some other combinations that either need to mean something, or they need to error.

What would this do?

`./bashunit -a assert_same "foo" "eval the_command"`

The 2nd argument can just be a normal string. I think special strings shouldn't switch the behaviour of the command.

My preferred syntax would be to have:

`bashunit -e "the_command"`

It'd mean that bashunit would run the command and then assert what we ask it to assert. Right now the only assertion would be exit code:

`bashunit -a exit_code "1" -e "the_command"`

Maybe other asserts could work on the output automatically?

`bashunit -a exit_code "1" -a contains "some error" -e "the_command"` (I don't know if I can currently pass multiple `-a` args to perform multiple assertions)

Also it'd be nice to save the output to a variable with another option for further asserts:

`bashunit -a exit_code "1" -e "the_command" -o $OUTPUT`

**Comment by Chemaclass** [comment]
I like your proposal of removing eval from the second string passed to the exit_code:
- `bashunit -a exit_code "1"` - compare with the last exit code
- `bashunit -a exit_code "1" "the_command"` - run the command inside bashunit and compare the codes
- `bashunit -a exit_code "1" "$(the_command)"` - run the command outside and compare the codes

Basically, the change I would suggest/do is keep the same behaviour (no need to introduce -e or -o) and only for the exit_code, if the second argument is a string then treat it as a callable (which will run internally with native bash `eval`). How does this sounds like, @ondrejmirtes ?

**Pull request merged**